WTG.Analyzers/Analyzers/StringBuilderAppend/StringBuilderAppendAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+using System.Collections.Immutable;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using WTG.Analyzers.Utils;
+
+namespace WTG.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public sealed class StringBuilderAppendAnalyzer : DiagnosticAnalyzer
+	{
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+			Rules.DontMutateAppendedStringArgumentsRule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction(Start);
+		}
+
+		static void Start(CompilationStartAnalysisContext context)
+		{
+			var cache = new FileDetailCache();
+
+			context.RegisterSyntaxNodeAction(c => Analyze(c, cache), SyntaxKind.InvocationExpression);
+		}
+
+		static void Analyze(SyntaxNodeAnalysisContext context, FileDetailCache cache)
+		{
+			if (cache.IsGenerated(context.SemanticModel.SyntaxTree, context.CancellationToken))
+			{
+				return;
+			}
+
+			var invokeExpression = (InvocationExpressionSyntax)context.Node;
+
+			if (!LooksLikeAppendMethod(invokeExpression) && invokeExpression.ArgumentList.Arguments.Count == 0)
+			{
+				return;
+			}
+
+			var firstArgument = invokeExpression.ArgumentList.Arguments[0].Expression;
+
+			if (!LooksLikeMutation(firstArgument))
+			{
+				return;
+			}
+
+			if (!IsAppendMethod(context.SemanticModel, invokeExpression, context.CancellationToken) ||
+				!IsMutation(context.SemanticModel, firstArgument, context.CancellationToken))
+			{
+				return;
+			}
+
+			context.ReportDiagnostic(Diagnostic.Create(
+				Rules.DontMutateAppendedStringArgumentsRule,
+				firstArgument.GetLocation(),
+				additionalLocations: new[] { invokeExpression.GetLocation() }));
+		}
+
+		static bool LooksLikeAppendMethod(InvocationExpressionSyntax invoke)
+			=> invoke.Expression.IsKind(SyntaxKind.SimpleMemberAccessExpression)
+				&& invoke.Expression is MemberAccessExpressionSyntax member
```

Comment:
Can we cast here instead of type-checking?

Reply:
By 'type-checking' I assume you are referring to the use of the `is` operator rather than the `IsKind` method.

We definitely need at least one of them and I prefer to keep the `IsKind` over the `is`. The only reason I chose to use `is` rather than a regular cast is so that it collapses nicely into an expression. I suppose I could have done something like:

```csharp
&& (MemberAccessExpressionSyntax)invoke.Expression is var member
```

but that feels kinda odd and I don't think it's any better.

Reply:
I don't remember what version of C# we're targeting here but another option I think as of C# 9 would be:

```csharp
invoke.Expression.IsKind(SyntaxKind.SimpleMemberAccessExpression)
&& (MemberAccessExpressionSyntax)invoke.Expression is {
    Name: { Identifier: { Text: nameof(Append) or nameof(AppendLine) }}
}
```

though that also introduces (unnecessary?) null-checks, but it also skips the `isinst` instruction and doesn't access each property twice.

So many different ways to do the same thing...

Reply:
We use [C#8](https://github.com/WiseTechGlobal/WTG.Analyzers/blob/89fb140f0dd62d0356917dffc8c6cb8530c7842e/Directory.Build.props#L8). I'm not particularly fond of that pattern matching approach, it looks messy to me.